entrance_rando.py

```diff
@@ -345,18 +345,18 @@ def find_pairing(dead_end: bool, require_new_regions: bool) -> bool:
 
     # stage 1 - try to place all the non-dead-end entrances
     while entrance_lookup.others:
-        if not find_pairing(False, True):
+        if not find_pairing(dead_end=False, require_new_regions=True):
             break
     # stage 2 - try to place all the dead-end entrances
     while entrance_lookup.dead_ends:
-        if not find_pairing(True, True):
+        if not find_pairing(dead_end=True, require_new_regions=True):
             break
     # stage 3 - connect any dangling entrances that remain
     while entrance_lookup.others:
-        find_pairing(False, False)
+        find_pairing(dead_end=False, require_new_regions=False)
     # stage 4 - last chance for dead ends
     while entrance_lookup.dead_ends:
-        find_pairing(True, False)
+        find_pairing(dead_end=True, require_new_regions=False)
```

Comment:
Stage 3 is to connect any dangling entrances, and stage 4 is the "last chance for dead ends". What's the consideration here?
I'd assume that any dead ends would be connected in stage 2, and that after stage 3 there should be no available entrances left, unless someone marked an area as a dead end that isn't actually a dead end.

Reply:
So, why don't stage 1 and 2 guarantee access to all regions? Ideally, those event-gated ones would not stop it from pairing all of the scenes. If the latch/switch/whatever you wanna call it is in a dead end, that dead end shouldn't be called a dead end so that it can get paired in the first go-around.

Reply:
I am curious about this, in particular the handling of dead ends in how the map itself is generated. Granted I don't personally understand all the inner workings of AP, and haven't fully read the logic, but from what does seem clear to me and coming from a graph generation perspective I do have some thoughts.

The first is, why not pair dead ends (regions with a single entrance/exit) with regions that are not dead ends first? The reason being that given the list of regions that have only 1 entrance (or those with entrances but no exits), if you always pair one with another region that is not a dead end then all your dead ends should get paired.

The second being, assuming you do the first, the dead end list can update with regions that were other (not dead ends) but can now be considered dead ends, for the purpose of generation. That is to say if you have room X which is a dead end, and room Y which is not dead end but only has 2 entrance exits, once you pair X to one of the entrances with Y, Y can now be thought of as a dead end (it's a region that is now itself and X).

If you follow this logic, you should end up in a situation where, after you process all the dead ends (and regions that "become" dead ends), you only have others which you can hook up as normal (making them dead ends where possible). Or you have no regions that are not dead ends, in which case there should only be 2 dead ends which could then be connected (I'd need to check this, but if more then you'd just connect as normal w/o creating orphans). This should also guarantee a connected graph.

There are of course multiple ways to do generation, the general logic is similar, but does imply 3 states not 4. You have dead ends and others, you have only others, or you just have dead ends.

(The require new regions might muddle this, not sure, but it seems like that would create less variety?)

Also feel free to completely ignore my thoughts if it doesn't fit or ask questions if I don't make sense.

Reply:
> The first is, why not pair dead ends (regions with a single entrance/exit) with regions that are not dead ends first? The reason being that given the list of regions that have only 1 entrance (or those with entrances but no exits), if you always pair one with another region that is not a dead end then all your dead ends should get paired.

The big reason this doesn't work is because of Entrance access logic which (as outlined in the documentation) is used to ensure to ensure a graph is generated which will not make the chance of fill errors notably higher than in a non-ER context. For this to work you would either have to limit the pairings only to exits which require only randomized items (potentially extremely difficult to check) or just ignore logic and roll the dice and hope that no transition ends of being directly or transitively self-locking

Reply:
The rest I think will be best if I sit and reread my code as it has now been months since I wrote it, and I can't remember all the details of implementation.

Reply:
Ah, here's the consideration for stage 4. Stages 1 and 2 get to pass if every remaining target points to an already-placed region. So if there is a dead end region with 2 one-way entrances but no exits, for example, that would need stage 4 to connect the second edge in.

In an ideal world stages 3 and 4 do not give any preference order to dead ends vs non dead ends, as both are serving the same purpose of trying to interconnect already-available regions. Maybe instead of dead-end being boolean it should be a flag enum going into group lookup to achieve that but I think the effect is negligible because this edge case is probably so rare and the current setup allows a good deal of code reuse.

Reply:
Actually this made me realize an issue with the logic for determining deadendedness, namely that a 2-way into/out of a region is a dead end in coupled but is not a dead end in uncoupled

Reply:
I didn't even pay attention to that, but yeah I knew exactly which part you were talking about when you just mentioned it. [this below](https://github.com/ArchipelagoMW/Archipelago/pull/2883#discussion_r1618221868)

I do think that in uncoupled mode dead ends would be those regions that have an entrance but the exit has already been used. I think this is partly why it was odd to me that dead ends weren't tracked beyond the initial entrance lookup creation. (whether this actually matters for this generation is unknown to me)

Reply:
The definition of a dead end is a target which cannot provide access to new randomized exits even when ignoring all logic conditions (this is also specified in the documentation). So you never need or want to recompute it. Here are some illustrative cases which I believe are not worth putting in the docs

```mermaid
graph LR
  x1:::hidden-->|e1| A
  x2:::hidden-->|e2| A
  classDef hidden display:none;
```
Both entrances to A (e1 and e2) are always dead ends because there is no way out, and therefore placing it may cap off the region graph mid generation (this is why we care about dead ends, we want to make sure that we don't close all our routes to expand too early).

---

```mermaid
graph LR
  x1:::hidden<-->|e1| A
  classDef hidden display:none;
```
The entrance to A (e1) is a dead end in uncoupled mode, because once you get in, you can go back out and end up elsewhere. It is a dead end in coupled mode as the reverse exit (also called e1 because this is the only way we have to do coupling) gets auto-placed to a region you have been already.

---

```mermaid
graph LR
  x1:::hidden-->|e1| A -->|e2| x2:::hidden
  classDef hidden display:none;
```
The entrance to A (e1) is never a dead end because it always can give access to the exit of A (e2)

Reply:
I ended up adjusting the comments and docs around this as well in the latest commit, when I switched the order to prevent ER from hardlocking